tests/client/button-click.test.js

```diff
@@ -0,0 +1,125 @@
+import { addWidget, widgetFilter } from '../../client/js/serverstate.js';
+import { Button } from '../../client/js/widgets/button.js';
+import { Widget } from '../../client/js/widgets/widget.js';
+
+import { addButton, addLabel, removeWidget } from './client-util.js';
+
+function createClickThisWidgets(testName, numWidgets, numWithClickThis) {
+  const buttonDef = {
+    "clickRoutine": [
+      {
+        "func": "LABEL",
+        "label": `${testName}-test-label`,
+        "value": 1,
+        "mode": "inc"
+      }
+    ],
+    "debug": false,
+    "type": "button"
+  }
+
+  const widgets = [];
+  let num = 0
+  while (num++ < numWidgets) {
+    buttonDef.id = `${testName}-button-${num}`;
+    buttonDef.clickThis = num <= numWithClickThis ? true : false;
+    let b = addButton(buttonDef);
+    widgets.push(b);
+  }
+  return widgets;
+}
+
+
+describe("Scenarios: Clicking widgets", () => {
+  const testName = "button-click";
+  let testButton;
+  let testLabel;
+  beforeAll(() => {
+    const testButtonDef = {
+      id: `${testName}-test-button`,
+      debug: false,
+      type: "button"
+    }
+    testButton = addButton(testButtonDef);
+
+    testLabel = addLabel(`${testName}-test-label`);
+  });
+  afterAll(() => {
+    removeWidget(testButton.p('id'));
+    removeWidget(testLabel.p('id'));
+  });
+
+  describe("Given a button that clicks widgets from an undefined collection", () => {
+    beforeAll(() => {
+      testButton.p('clickRoutine', [
+        {
+          "func": "CLICK",
+          "collection": "undefined"
+        }
+      ]);
+    });
+    describe("When clicked", () => {
+      test("Then it does not throw an error", () => {
+        expect(async () => {
```

Comment:
Did you test this `async` construct if the function actually _does_ throw something?

Looks like it's used everywhere so I'll assume the answer is "yes".

Reply:
It's async because of the `await` on click() because click() is an `async` function. I'm pretty sure it will fail if `async` isn't included here.

Reply:
Yes but I'm not sure this works as expected. If I add `throw "TEST";` into `Button.click` below `if(a.func == 'CLICK')`, I'm pretty sure I see Node.js crashing instead of it telling me that the test condition failed. Not sure if that's expected behaviour though.

Reply:
No, it shouldn't crash. That's a good point. I'll update it so the thrown error gets caught and the test marked fail instead of crashing.

Reply:
Fixed. Also added a handler to rethrow uncaught promise rejections so that the full stack trace will be provided on the console.

To observe the behavior of the adjusted test code, disable the `if(isValidCollection(...` line in button.js. The failure then looks like this:

![image](https://user-images.githubusercontent.com/10050028/103966847-adbb2280-512e-11eb-89f1-72e435b1df70.png)